### ISSUE TITLE
Pageable Design Flaws

### ISSUE THREAD

**Issue by amihaiemil** [spring-projects/spring-data-commons#2674]
I realize this is up for discussion and not easily changeable, except through long-time deprecation, but here's two big problems with it:

**1. Pageable has pageNumber and pageSize.**

The concept of "page" doesn't actually exist. A page is the consquence of index (number of elements to skip reading) and limit (pageSize). So it doesn't really make sense to ever have "pageNumber" anywhere. The only place where page number makes sense is the UI, in my opinion. Parameters index (or start) and size (or limit, or pageSize, as it is) are the correct ones.

Let's say this is debatable, but the second:

**2. Pageable.unpaged() has a bad design**

Because it throws exceptions instead of, for example, pageNumber 0 and pageSize MAX_INT. I am using unpaged as a client to a custom repository method, and I get these exceptions, because the repository tries to pass pageNumber and pageSize to a Mongo Aggrengation for example. 

Even more, Pageable.unpaged() is a bad encouragement to users. If I have a repository method which accepts Pageable, it is clear that that method can return *a lot* of records.

Why allow the user to give me an "unpaged" pageable? There is no scenario in which a ton of records should be returned, in my opinion. Let's say the user can implement their own "unpaged" Pageable anyway, but why give them this utility method? It encourages bad practice. 

Thank you for reading and your consideration.

**Comment by jxblum** [comment]
#1 is definitely debatable.  Given a Page Size and Page Number, then you can compute an index into any ordered, or indexable data structure.

**Comment by amihaiemil** [comment]
@jxblum yep, but the page number is the thing which should be calculated, because the page is the abstract concept here, it doesn't actually exist. Page number is derived from index and size, imo. 

This is how I explain it to juniors who are learning pagination. Because they start thinking about page number and page size, then they get confused, particularly about what happens when a page is skipped? Or when a user says page 1 with 5 records per pege, then says page 2 and 7 records per page? All this goes away when you work with index and size, rather than page and size.

**Comment by jxblum** [comment]
Well, that is the point actually... 

The `Pageable` interface in _Spring Data Commons_ is an abstraction around any data structure (supported by the underlying data store), that is orderable / indexed or NOT.

You could imagine paging a non-indexed, non-orderable data structure as well. This is entirely possible and in some cases (e.g. NoSQL data stores) quite necessary. You could argue why would anyone do that, or it seems rather pointless.  But in fact, even the order and/or number of elements in a ordered, indexable data structure matching some predicate could change between 1 request (search) from another.

Additionally, the "order" (sort) may be determined on the fly, at time of request, depending on ordering criteria and/or data structure returned from the the underlying store, which may be unordered.

Also, I don't agree with this:

> _All this goes away when you work with index and size_

**Comment by amihaiemil** [comment]
@jxblum maybe I am missing something, but what sense does it make to try to paginate unstructured/unordered data? pageNumber or index approach, it doesn't matter, records could just repear between the pages or not show up at all, maybe even going in an endless loop?

What am I missing?

**Comment by jxblum** [comment]
A `Set is "_an unordered collection of elements_", but the `Set` can still be ordered, on-the-fly if necessary.

In addition, the modification of elements or insertion/deletion of elements can occur between requests (searches), even in an ordered (indexable) data structure thus changing the pages structure/contents, assuming the backend data store is queried on each request/search in a highly concurrent system. So, in other words, though pagination implies an order, it does not necessarily require the elements to be ordered. 

For example, perhaps you are applying pagination of a stream of elements. Get the first N number of elements in a stream, then get the next M elements in a stream, and so on. This is not entirely unlike Reactive Streams. And, while you can generally progress forward in a stream, you may not necessarily be able to go back.

So, in summary, it depends.

Also, keep in mind that what is generally in Spring Data Commons is a general, LCD, "abstraction" supporting a variety of dat stores.

**Comment by amihaiemil** [comment]
@jxblum I see. Well, I would leave this up for discussion, I am not sure what the process is, how long can it stay open, how many opinions are taken into consideration etc 😅

**Comment by jxblum** [comment]
Sorry for the delayed response, I was traveling yesterday. We typically triage and discuss these sort of issues as a team to prioritize them and determine the correct course of action.

Ultimately, I am not sure what the outcome of this will be as it is quite subjective, or debatable as you mention above. 

At any rate, thank you for the feedback and lively discussion. It is much appreciated. Cheers!

**Comment by amihaiemil** [comment]
@jxblum thanks! Do let me know of the outcome, please. 😁

**Comment by jxblum** [comment]
Certainly, we will post the outcome in this ticket at some future point.

**Comment by mp911de** [comment]
Abstract concepts cannot please everyone. Our concept of a page of objects (which is already difficult to translate into the real world) is modeled along offset/limit functionality encapsulated through the `Pageable` type hierarchy.

The surprising behavior of `Pageable.unpaged()` is indeed not documented; we fixed that with #2679. `unpaged()` follows the design of `Optional` instead of using `null` in places where we do not have sorting or paging information.

**Comment by amihaiemil** [comment]
@mp911de Thank you for your response and JavaDoc improvement!
